nixos/modules/services/networking/nftables.nix

```diff
@@ -124,25 +229,58 @@ in
     boot.blacklistedKernelModules = [ "ip_tables" ];
     environment.systemPackages = [ pkgs.nftables ];
     networking.networkmanager.firewallBackend = mkDefault "nftables";
+    # versionOlder for backportability, remove afterwards
+    networking.nftables.flushRuleset = mkDefault (versionOlder config.system.stateVersion "23.05" || (cfg.rulesetFile != null || cfg.ruleset != ""));
     systemd.services.nftables = {
       description = "nftables firewall";
       before = [ "network-pre.target" ];
       wants = [ "network-pre.target" ];
       wantedBy = [ "multi-user.target" ];
       reloadIfChanged = true;
       serviceConfig = let
+        enabledTables = filterAttrs (_: table: table.enable) cfg.tables;
+        deletionsScript = pkgs.writeScript "nftables-deletions" ''
+          #! ${pkgs.nftables}/bin/nft -f
+          ${if cfg.flushRuleset then "flush ruleset"
+            else concatStringsSep "\n" (mapAttrsToList (_: table: ''
+              table ${table.family} ${table.name}
+              delete table ${table.family} ${table.name}
+            '') enabledTables)}
+          ${cfg.extraDeletions}
+        '';
+        deletionsScriptVar = "/var/lib/nftables/deletions.nft";
+        ensureDeletions = pkgs.writeShellScript "nftables-ensure-deletions" ''
+          touch ${deletionsScriptVar}
+          chmod +x ${deletionsScriptVar}
+        '';
+        saveDeletionsScript = pkgs.writeShellScript "nftables-save-deletions" ''
+          cp ${deletionsScript} ${deletionsScriptVar}
+        '';
+        cleanupDeletionsScript = pkgs.writeShellScript "nftables-cleanup-deletions" ''
+          rm ${deletionsScriptVar}
+        '';
         rulesScript = pkgs.writeTextFile {
           name =  "nftables-rules";
           executable = true;
           text = ''
             #! ${pkgs.nftables}/bin/nft -f
-            flush ruleset
-            ${if cfg.rulesetFile != null then ''
+            # previous deletions, if any
+            include "${deletionsScriptVar}"
```

Comment:
after applying this patch with a nixos-rebuild switch, nfttables.service failed with following error:
Apr 22 00:36:33 xx systemd[1]: Reloading nftables firewall...
Apr 22 00:36:33 xx gfzcn50sarqv9s711n1zfy6yky1wa2rf-nftables-rules[1274297]: /nix/store/gfzcn50sarqv9s711n1zfy6yky1wa2rf-nftables-rules:3:1-42: Error: File not found: /var/lib/nftables/deletions.nft
Apr 22 00:36:33 xx gfzcn50sarqv9s711n1zfy6yky1wa2rf-nftables-rules[1274297]: include "/var/lib/nftables/deletions.nft"
Apr 22 00:36:33 xx gfzcn50sarqv9s711n1zfy6yky1wa2rf-nftables-rules[1274297]: ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

I have to manually restart nftables.service again to work. Maybe we missed a check for deletionsScriptVar here?

Reply:
yes, we missed an `ensureDeletions` in `ExecReload`